project/src/pv_circular_buffer.c

```diff
@@ -0,0 +1,170 @@
+    }
+    if (element_size <= 0) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if (!object) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+
+    *object = NULL;
+
+    pv_circular_buffer_t *o = calloc(1, sizeof(pv_circular_buffer_t));
+    if (!o) {
+        return PV_CIRCULAR_BUFFER_STATUS_OUT_OF_MEMORY;
+    }
+
+    o->buffer = malloc(element_count * element_size);
+    if (!(o->buffer)) {
+        pv_circular_buffer_delete(o);
+        return PV_CIRCULAR_BUFFER_STATUS_OUT_OF_MEMORY;
+    }
+
+    o->capacity = element_count;
+    o->element_size = element_size;
+
+    *object = o;
+
+    return PV_CIRCULAR_BUFFER_STATUS_SUCCESS;
+}
+
+void pv_circular_buffer_delete(pv_circular_buffer_t *object) {
+    if (object) {
+        free(object->buffer);
+        free(object);
+    }
+}
+
+pv_circular_buffer_status_t pv_circular_buffer_read(
+        pv_circular_buffer_t *object,
+        void *buffer,
+        int32_t buffer_length) {
+    if (!object) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if (!buffer) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if (!buffer_length) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if ((buffer_length <= 0) || (buffer_length > object->capacity)) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+
+    void *dst_ptr = buffer;
+    const void *src_ptr = (char *) object->buffer + (object->read_index * object->element_size);
+
+    const int32_t available = object->capacity - object->read_index;
+    const int32_t max_copy = (object->count < buffer_length) ? object->count : buffer_length;
+    const int32_t to_copy = (max_copy < available) ? max_copy : available;
+
+    memcpy(dst_ptr, src_ptr, to_copy * object->element_size);
+
+    object->read_index = (object->read_index + to_copy) % object->capacity;
+
+    const int32_t remaining = max_copy - to_copy;
+    if (remaining > 0) {
+        dst_ptr = (char *) buffer + (to_copy * object->element_size);
+        src_ptr = object->buffer;
+
+        memcpy(dst_ptr, src_ptr, remaining * object->element_size);
+
+        object->read_index = remaining;
+    }
+
+    object->count -= max_copy;
+
+    return PV_CIRCULAR_BUFFER_STATUS_SUCCESS;
+}
+
+pv_circular_buffer_status_t pv_circular_buffer_write(
+        pv_circular_buffer_t *object,
+        const void *buffer,
+        int32_t buffer_length) {
+    if (!object) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if (!buffer) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if ((buffer_length <= 0) || (buffer_length > object->capacity)) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+
+    const int32_t available = object->capacity - object->write_index;
+    const int32_t to_copy = (buffer_length < available) ? buffer_length : available;
```

Comment:
what would happen to the rest of the input buffer? it seems that they're ignored, right?

Reply:
I think you're right, thanks for the catch, I pushed up a fix